fms/utils/generation.py

```diff
@@ -74,9 +180,33 @@ def generate(
     kwargs: MutableMapping[str, Any] = dict()
     kwargs["past_key_value_states"] = None
     kwargs["use_cache"] = use_cache
+    kwargs["position_ids"] = position_ids
 
-    for _ in range(max_new_tokens):
+    for i in range(max_new_tokens):
         input_ids = next_input[:, -max_seq_len:]
+
```

Comment:
It still seems like we've got a lot of code here, much of it confusing, that complicates the generate function. I think we should try to keep generate pretty minimally just doing one thing -- the generation loop. You've got some helper functions that move some of the code out, but I think this could be simplified further (both here and above).

We already handle batched tensor inputs. The only changes we're trying to make here are to:
* convert a list of tensors input_ids into a single padded batched tensor.
* add a mask and position_ids to the kwargs in the case where the input_ids were a list of tensors ( length < 1).

I don't think those two things should add 50+ LOC to the generate function, maybe they could each be a single function call?